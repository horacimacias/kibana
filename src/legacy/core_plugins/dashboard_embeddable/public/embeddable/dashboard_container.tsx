/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import uuid from 'uuid';

import { I18nProvider } from '@kbn/i18n/react';
import { IndexPattern } from 'ui/index_patterns';

import { TimeRange } from 'ui/timefilter/time_history';
import {
  Container,
  ContainerInput,
  Embeddable,
  EmbeddableFactoryRegistry,
  EmbeddableInput,
  Filter,
  Query,
  RefreshConfig,
  ViewMode,
  EmbeddableOutput,
  EmbeddableInputMissingFromContainer,
  isErrorEmbeddable,
  EmbeddableFactory,
} from '../../../embeddable_api/public/index';

import { DASHBOARD_CONTAINER_TYPE } from './dashboard_container_factory';
import { createPanelState } from './panel';
import { DashboardPanelState } from './types';
import { DashboardViewport } from './viewport/dashboard_viewport';

export interface DashboardContainerInput extends ContainerInput {
  viewMode: ViewMode;
  filters: Filter[];
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  expandedPanelId?: string;
  useMargins: boolean;
  title: string;
  description?: string;
  isFullScreenMode: boolean;
  panels: { [panelId: string]: DashboardPanelState };

  // Used to force a refresh of embeddables even if there were no other input state
  // changes.
  lastReloadRequestTime?: number;
}

export interface DashboardEmbeddableInput {
  filters: Filter[];
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  viewMode: ViewMode;
  hidePanelTitles?: boolean;
  id: string;
  firstName: string;
}

export interface DashboardEmbeddableOutput extends EmbeddableOutput {
  indexPatterns?: IndexPattern[];
}

export type DashboardEmbeddable = Embeddable<
  DashboardEmbeddableInput & EmbeddableInput,
  DashboardEmbeddableOutput
>;

export class DashboardContainer extends Container<
  DashboardEmbeddableInput,
  DashboardEmbeddableOutput,
  DashboardContainerInput
> {
  constructor(
    initialInput: DashboardContainerInput,
    embeddableFactories: EmbeddableFactoryRegistry,
    parent?: Container
  ) {
    super(
      DASHBOARD_CONTAINER_TYPE,
      {
        panels: {},
        isFullScreenMode: false,
        filters: [],
        useMargins: true,
        ...initialInput,
      },
      { embeddableLoaded: {} },
      embeddableFactories,
      parent
    );
  }

  protected createNewPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    factory: EmbeddableFactory<EEI>,
    partial: Partial<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>> & {
      id?: string;
    } = {}
  ): DashboardPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>> {
    const panelState = super.createNewPanelState(factory, partial);
    return createPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>>(
      panelState,
      Object.values(this.input.panels)
    );
  }

  public onToggleExpandPanel(id: string) {
    const newValue = this.input.expandedPanelId ? undefined : id;
    this.updateInput({
      expandedPanelId: newValue,
    });
  }

  public onPanelsUpdated = (panels: { [panelId: string]: DashboardPanelState }) => {
    this.updateInput({
      panels: {
        ...panels,
      },
    });
  };

  public onExitFullScreenMode = () => {
    this.updateInput({
      isFullScreenMode: false,
    });
  };

  public render(dom: React.ReactNode) {
    ReactDOM.render(
      // @ts-ignore
      <I18nProvider>
        <DashboardViewport embeddableFactories={this.embeddableFactories} container={this} />
      </I18nProvider>,
      dom
    );
  }

  public getPanelIndexPatterns() {
    const indexPatterns: IndexPattern[] = [];
    Object.values(this.embeddables).forEach(embeddable => {
      if (!isErrorEmbeddable(embeddable)) {
        const embeddableIndexPatterns = embeddable.getOutput().indexPatterns;
        if (embeddableIndexPatterns) {
          indexPatterns.push(...embeddableIndexPatterns);
        }
      }
    });
    return indexPatterns;
  }

  protected getInheritedInput(id: string): DashboardEmbeddableInput {
    const { viewMode, refreshConfig, timeRange, query, hidePanelTitles, filters } = this.input;
    return {
      filters,
      hidePanelTitles,
      query,
      timeRange,
      refreshConfig,
      viewMode,
      id,
      firstName: 'suuuue',
    };
  }
}
