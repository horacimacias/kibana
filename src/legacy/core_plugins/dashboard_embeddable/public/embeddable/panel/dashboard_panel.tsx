/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import _ from 'lodash';
import React from 'react';

import { EuiLoadingChart } from '@elastic/eui';
import { InjectedIntl, injectI18n } from '@kbn/i18n/react';

import { ErrorEmbeddable } from 'plugins/embeddable_api/index';

import { DashboardContainer, DashboardEmbeddable } from '../dashboard_container';

export interface DashboardPanelProps {
  lastReloadRequestTime?: number;
  embeddableId: string;
  className?: string;
  container: DashboardContainer;
}

export interface DashboardPanelUiProps extends DashboardPanelProps {
  intl: InjectedIntl;
}

interface State {
  loading: boolean;
}

export function isErrorEmbeddable(
  embeddable: ErrorEmbeddable | DashboardEmbeddable | DashboardContainer
): embeddable is ErrorEmbeddable {
  return (embeddable as ErrorEmbeddable).getInput().errorMessage !== undefined;
}

class DashboardPanelUi extends React.Component<DashboardPanelUiProps, State> {
  [panel: string]: any;
  public mounted: boolean;
  public embeddable!: DashboardEmbeddable | ErrorEmbeddable;
  private unsubscribe?: () => void;

  constructor(props: DashboardPanelUiProps) {
    super(props);
    this.state = {
      loading: true,
    };

    this.mounted = false;
  }

  public async componentDidMount() {
    this.mounted = true;
    if (!this.embeddable) {
      this.renderEmbeddable();
    }
    this.unsubscribe = this.props.container.subscribeToOutputChanges(() => {
      if (!this.embeddable) {
        this.renderEmbeddable();
      }
    });
  }

  public componentWillUnmount() {
    this.mounted = false;
    if (
      this.embeddable &&
      !isErrorEmbeddable(this.embeddable) &&
      this.props.container.getEmbeddable(this.embeddable.id)
    ) {
      this.props.container.removeEmbeddable(this.embeddable);
    }
    if (this.unsubscribe) {
      this.unsubscribe();
    }
  }

  public renderEmbeddableViewport() {
    const classes = classNames('embPanel panel-content', {
      'panel-content-isLoading': this.state.loading,
    });

    return (
      <div
        id="embeddedPanel"
        className={classes}
        ref={panelElement => (this.panelElement = panelElement)}
      >
        {this.state.loading && <EuiLoadingChart size="l" mono />}
      </div>
    );
  }

  public shouldComponentUpdate(nextProps: DashboardPanelUiProps) {
    if (this.embeddable && nextProps.lastReloadRequestTime !== this.props.lastReloadRequestTime) {
      this.embeddable.reload();
    }

    return true;
  }

  public render() {
    return this.renderEmbeddableViewport();
  }

  private renderEmbeddable() {
    this.embeddable = this.props.container.getEmbeddable(this.props.embeddableId);
    if (this.mounted && this.embeddable) {
      this.setState({ loading: false }, () => {
        this.embeddable.renderInPanel(this.panelElement, this.props.container);
      });
    } else if (this.embeddable) {
      this.embeddable.destroy();
    }
  }
}

export const DashboardPanel = injectI18n(DashboardPanelUi);
