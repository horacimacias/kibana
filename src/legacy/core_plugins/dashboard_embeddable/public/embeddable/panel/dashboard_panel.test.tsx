/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

jest.mock('ui/metadata', () => ({
  metadata: {
    branch: 'my-metadata-branch',
    version: 'my-metadata-version',
  },
}));

import { EuiLoadingChart } from '@elastic/eui';
import {
  HelloWorldEmbeddable,
  HelloWorldInput,
} from 'plugins/embeddable_api/__test__/embeddables/hello_world_embeddable';
import { embeddableFactories } from 'plugins/embeddable_api/index';
import { QueryLanguageType, ViewMode } from 'plugins/embeddable_api/types';
import React from 'react';
import { mountWithIntl } from 'test_utils/enzyme_helpers';
import { DashboardContainer, DashboardContainerInput } from '../dashboard_container';
import { DashboardPanel } from './dashboard_panel';
import {
  HELLO_WORLD_EMBEDDABLE,
  FilterableEmbeddable,
  FilterableEmbeddableInput,
  FILTERABLE_EMBEDDABLE,
} from 'plugins/embeddable_api/__test__';

function getDashboardContainerInput(): DashboardContainerInput {
  return {
    id: '123',
    viewMode: ViewMode.EDIT,
    filters: [],
    query: {
      language: QueryLanguageType.KUERY,
      query: '',
    },
    timeRange: {
      to: 'now',
      from: 'now-15m',
    },
    useMargins: false,
    title: 'My Test Dashboard',
    isFullScreenMode: false,
    panels: {},
  };
}

test('DashboardPanel renders an embeddable when it is done loading', async () => {
  const container = new DashboardContainer(getDashboardContainerInput(), embeddableFactories);
  const newEmbeddable = await container.addNewEmbeddable<HelloWorldInput, HelloWorldEmbeddable>(
    HELLO_WORLD_EMBEDDABLE,
    {
      firstName: 'Sue',
      id: '123',
    }
  );

  expect(newEmbeddable.id).toBeDefined();

  const component = mountWithIntl(
    <DashboardPanel.WrappedComponent
      intl={null as any}
      container={container}
      embeddableId={newEmbeddable.id}
    />
  );

  expect(component).toMatchSnapshot();

  const loadingIndicator = component.find(EuiLoadingChart);
  expect(loadingIndicator).toHaveLength(0);
});
