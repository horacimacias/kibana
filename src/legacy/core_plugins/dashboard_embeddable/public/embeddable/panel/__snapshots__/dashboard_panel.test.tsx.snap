// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPanel renders an embeddable when it is done loading 1`] = `
<DashboardPanelUi
  container={
    DashboardContainer {
      "anyChangeListeners": Array [
        [Function],
        [Function],
      ],
      "destoyed": false,
      "embeddableFactories": EmbeddableFactoryRegistry {
        "factories": Object {
          "FILTERABLE_EMBEDDABLE": FilterableEmbeddableFactory {
            "name": "FILTERABLE_EMBEDDABLE",
            "savedObjectMetaData": undefined,
          },
          "dashboard": DashboardContainerFactory {
            "name": "dashboard",
            "savedObjectMetaData": undefined,
          },
          "hello_world": HelloWorldEmbeddableFactory {
            "name": "hello_world",
            "savedObjectMetaData": undefined,
          },
        },
      },
      "embeddables": Object {
        "123": HelloWorldEmbeddable {
          "anyChangeListeners": Array [],
          "destoyed": false,
          "id": "123",
          "input": Object {
            "filters": Array [],
            "firstName": "Sue",
            "hidePanelTitles": undefined,
            "id": "123",
            "query": Object {
              "language": "kuery",
              "query": "",
            },
            "refreshConfig": undefined,
            "timeRange": Object {
              "from": "now-15m",
              "to": "now",
            },
            "viewMode": "edit",
          },
          "inputChangeListeners": Array [
            [Function],
          ],
          "isContainer": false,
          "node": <div
            class="embeddable-root panel-content"
          >
            <div>
              Hello World!
              <br />
               Sincerly, $
              
            </div>
          </div>,
          "output": Object {
            "birthLastName": undefined,
            "birthName": "Sue",
            "name": "Sue",
          },
          "outputChangeListeners": Array [
            [Function],
          ],
          "panelContainer": <div
            class="embPanel panel-content"
            id="embeddedPanel"
          >
            <div
              class="euiPanel embPanel embPanel--editing"
              data-test-subj="embeddablePanel"
            >
              <div
                class="embPanel__header embPanel__header--floater"
                data-test-subj="dashboardPanelHeading-"
              >
                <div
                  aria-label="Dashboard panel: "
                  class="embPanel__title embPanel__dragger"
                  data-test-subj="dashboardPanelTitle"
                />
                <div
                  class="euiPopover euiPopover--anchorDownRight euiPopover--withTitle embPanel_optionsMenuPopover"
                  data-test-subj="embeddablePanelContextMenuClosed"
                  id="dashboardPanelContextMenu"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      aria-label="Panel options"
                      class="euiButtonIcon euiButtonIcon--text embPanel_optionsMenuButton"
                      data-test-subj="embeddablePanelToggleMenuIcon"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiButtonIcon__icon"
                        focusable="false"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                        xmlns:xlink="http://www.w3.org/1999/xlink"
                      >
                        <defs>
                          <path
                            d="M.164 10.329L1.87 8 .163 5.67c.18-.601.43-1.19.758-1.757a8.197 8.197 0 0 1 1.142-1.535l2.872.313L6.099.05a8.166 8.166 0 0 1 3.8-.003l1.166 2.644 2.872-.313a8.166 8.166 0 0 1 1.899 3.293L14.13 8l1.706 2.33c-.18.601-.43 1.19-.758 1.757a8.197 8.197 0 0 1-1.142 1.535l-2.872-.313-1.164 2.641a8.166 8.166 0 0 1-3.8.003l-1.166-2.644-2.872.313a8.166 8.166 0 0 1-1.899-3.293zm4.663 1.986a1 1 0 0 1 1.023.591l.957 2.17c.79.134 1.597.132 2.387-.001l.956-2.169a1 1 0 0 1 1.023-.59l2.358.256a7.23 7.23 0 0 0 1.194-2.068l-1.401-1.913a1 1 0 0 1 0-1.182l1.4-1.912a7.165 7.165 0 0 0-1.192-2.069l-2.359.257a1 1 0 0 1-1.023-.591L9.193.924a7.165 7.165 0 0 0-2.387.001L5.85 3.094a1 1 0 0 1-1.023.59l-2.358-.256a7.23 7.23 0 0 0-1.194 2.068l1.401 1.913a1 1 0 0 1 0 1.182l-1.4 1.912c.28.751.681 1.45 1.192 2.069l2.359-.257zM8 11a3 3 0 1 1 0-6 3 3 0 0 1 0 6zm0-1a2 2 0 1 0 0-4 2 2 0 0 0 0 4z"
                            id="gear-a"
                          />
                        </defs>
                        <use
                          xlink:href="#gear-a"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="embeddable-root panel-content"
              >
                <div>
                  Hello World!
                  <br />
                   Sincerly, $
                  
                </div>
              </div>
            </div>
          </div>,
          "parent": [Circular],
          "parentChangesUnsubscribe": [Function],
          "type": "hello_world",
          "unsubscribe": [Function],
        },
      },
      "id": "123",
      "input": Object {
        "filters": Array [],
        "id": "123",
        "isFullScreenMode": false,
        "panels": Object {
          "123": Object {
            "embeddableId": "123",
            "gridData": Object {
              "h": 15,
              "i": "123",
              "w": 24,
              "x": 0,
              "y": 0,
            },
            "partialInput": Object {
              "firstName": "Sue",
              "id": "123",
            },
            "type": "hello_world",
          },
        },
        "query": Object {
          "language": "kuery",
          "query": "",
        },
        "timeRange": Object {
          "from": "now-15m",
          "to": "now",
        },
        "title": "My Test Dashboard",
        "useMargins": false,
        "viewMode": "edit",
      },
      "inputChangeListeners": Array [],
      "isContainer": true,
      "onExitFullScreenMode": [Function],
      "onPanelsUpdated": [Function],
      "output": Object {
        "embeddableLoaded": Object {
          "123": true,
        },
      },
      "outputChangeListeners": Array [
        [Function],
      ],
      "panelContainer": undefined,
      "parent": undefined,
      "parentChangesUnsubscribe": undefined,
      "type": "dashboard",
    }
  }
  embeddableId="123"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {
        "date": Object {
          "full": Object {
            "day": "numeric",
            "month": "long",
            "weekday": "long",
            "year": "numeric",
          },
          "long": Object {
            "day": "numeric",
            "month": "long",
            "year": "numeric",
          },
          "medium": Object {
            "day": "numeric",
            "month": "short",
            "year": "numeric",
          },
          "short": Object {
            "day": "numeric",
            "month": "numeric",
            "year": "2-digit",
          },
        },
        "number": Object {
          "currency": Object {
            "style": "currency",
          },
          "percent": Object {
            "style": "percent",
          },
        },
        "relative": Object {
          "days": Object {
            "units": "day",
          },
          "hours": Object {
            "units": "hour",
          },
          "minutes": Object {
            "units": "minute",
          },
          "months": Object {
            "units": "month",
          },
          "seconds": Object {
            "units": "second",
          },
          "years": Object {
            "units": "year",
          },
        },
        "time": Object {
          "full": Object {
            "hour": "numeric",
            "minute": "numeric",
            "second": "numeric",
            "timeZoneName": "short",
          },
          "long": Object {
            "hour": "numeric",
            "minute": "numeric",
            "second": "numeric",
            "timeZoneName": "short",
          },
          "medium": Object {
            "hour": "numeric",
            "minute": "numeric",
            "second": "numeric",
          },
          "short": Object {
            "hour": "numeric",
            "minute": "numeric",
          },
        },
      },
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": Symbol(react.fragment),
      "timeZone": null,
    }
  }
>
  <div
    className="embPanel panel-content"
    id="embeddedPanel"
  />
</DashboardPanelUi>
`;
